### ISSUE TITLE
Why no symlink to python3

### ISSUE THREAD

**Issue by madhavajay** [iusrepo/python36u#2]
I understand you might want to have multiple versions installed, but by default would it not be useful if a symlink from /usr/bin/python3 to the binary was created in the event there is no symlink?

**Comment by b-harper** [comment]
Hey @madhavajay,

Thanks for reaching out.  Starting with our python35u packages, we switched from only allowing a single version of python 3 to allowing them to be installed in parallel.  Part of this process was to remove the symlink for /usr/bin/python3.  

We have had discussions regarding the best way to handle this issue.  All kinds of idea came up from something like what you proposed to even the use of ```alternatives```.  There was no clear cut 'best choice'.  At that time it was decided that leaving it up to the end user was the best option.

**Comment by carlwgeorge** [comment]
Here is the [relevant discussion](https://github.com/iuscommunity/wishlist/issues/17#issuecomment-146661217) from when we implemented this.

**Comment by madhavajay** [comment]
Hi everyone, yes I assumed that someone had made a conscious decision about this. Does Yum allow params to be passed to it when installing a package? I wonder if that would allow for a:
```
$ yum install python36u --symlink
```
Or something to that affect.
The only reason I bring it up is im using this awesome IUS (thanks guys) repo / package to provide python3 to my Centos 7 boxes with Ansible.

Without the symlink, I have a separate step which adds the symlink but then this is also brittle as its specific to the binary filename version.

Anyway, thanks for the quick response.
If there is the ability to take command line args and someone else thinks its also a good idea, i'd be happy to provide a PR if someone can point me in the right direction on which file to modify.

Iv not worked on an RPM package installer script before but i'm sure its fun. 👍

**Comment by carlwgeorge** [comment]
> Does Yum allow params to be passed to it when installing a package?

Rpmbuild supports `--with`/`--without` flags, but that is only for building the rpm.  Yum has no such feature.

> Without the symlink, I have a separate step which adds the symlink but then this is also brittle as its specific to the binary filename version.

The package name is also specific to the version, so how would that be brittle?  python35u will always have `/usr/bin/python3.5`, python36u will always have `/usr/bin/python3.6`, and so on.

**Comment by madhavajay** [comment]
If someone bumps the yum package version and forgets to bump the code which creates the symlink it will break.

Hrm, pity theres no flags for yum, what about an ENV variable?

**Comment by carlwgeorge** [comment]
No, the symlink will not break in that scenario.  If you create a symlink for `/usr/bin/python3 -> /usr/bin/python3.6`, and python36u gets updated from 3.6.2 to 3.6.3, the symlink will work exactly the same before and after the update.

There isn't any yum flag or environment variables that can do what you are looking for.  There just isn't a way to have multiple python3?u packages containing `/usr/bin/python3` symlinks to different locations.

**Comment by madhavajay** [comment]
Right, but there will be python3.7 and python3.8 and so on.
Anyway, yes I can see that there isnt really any optimal way to handle this, pity.

Thanks for the help explaining it.

**Comment by rahvee** [comment]
Given that python is backward compatible (i.e. python 3.5 can be run by python 3.6 and so on), I strongly disagree with the conclusion that "there isn't an optimal way to handle this." By eliminating the python3 symlink, you require everyone to hard-code python versions into their scripts, or manually create symlinks on every system. The de-facto standard is to use "`#!/usr/bin/env python3`" to execute whatever python3 happens to be on the `$PATH`. Just google for "shebang python" and see what I mean. It's universal, everywhere. `#!/usr/bin/env python` for python2, and `#!/usr/bin/env python3` for python 3.

**Comment by rahvee** [comment]
In response to @carlwgeorge , if you can't have a `/usr/bin/python3` symlink in the rpm, you can use a post-install script to link to whatever is the highest number python3 that's available.

**Comment by carlwgeorge** [comment]
@rahvee Thank you for sharing your thoughts on the matter.  We considered these concerns during the original discussion when we first implemented this change in python35u.  We decided that the concerns were minor enough to proceed with making our python3 packages fully parallel installable.

We still feel there isn't an optimal way to handle a `/usr/bin/python3` symlink in our packages.  Imagine if we chose the stance of having the latest version we ship contain the symlink.  Once python37u comes out, we would have to remove the symlink from python36u.  This would cause problems for any users of python36u depending on `/usr/bin/python3` to exist, and we would have no way to ensure they also had python37u installed.  Our stance is it's much better for users to face the problem immediately upon installation, either by using a more specific shebang line or creating their own symlink outside of the package.

Creating a symlink during %post would be error prone and difficult to maintain, and could also have unintended user impact.  A user may be primarily using python36u, but want to install python37u to see if their code is compatible.  Surprising them by switching `/usr/bin/python3` to python37u upon installation is not in line with the goals of IUS.  I understand that python's goal is to be backwards compatible, but if that were 100% true then upstream wouldn't be maintaining 3.4, 3.5, and 3.6 stable branches.

**Comment by rahvee** [comment]
Thank you @carlwgeorge I think there still exists a possible solution: If there were an additional rpm, which contains nothing but a symlink. The package could be called something like "python36u-symlink" and it could conflict with python35u-symlink, and it can be dependent on python36u that provides the target binary /usr/bin/python36. When somebody does "yum provides /usr/bin/python3" they'll be able to easily identify the list of available packages to choose from. They won't get surprise updated when they install python37u; they will have to remove python36u-symlink and install python37u-symlink in order to make python37u the new default. 

Right now, when somebody does `yum provides /usr/bin/python3` they would mistakenly be led to believe that python34u is the latest available. I am very, very certain that it's unexpected by a vast majority of users out there, that they need to manually create a symlink outside of their package manager. Even if it seems like more work to you or somebody else, to install and remove a package as opposed to manually creating a symlink, people who manage large scale infrastructures (as I do) are deploying packages via puppet, ansible, chef, etc. Yes we *can* also deploy symlinks that way, but when we're talking about updating packages, it's a big surprise that such a step is necessary, and when we're searching for what the latest version of something is, there's no intuitive way using standard techniques (such as yum provides) to learn that we need to do this unexpected requirement.

**Comment by carlwgeorge** [comment]
That is an interesting idea.  Since this discussion is not just for python36u, but rather all IUS python3 packages, I created a new issue summarizing it our [packaging meta repo](https://github.com/iuscommunity/packaging/issues/1) with the intent of gathering feedback from a wider audience.  Please continue the discussion with us in that thread.  I think I've summarized your thoughts so far, but go ahead and comment there if I left anything out.
